[libc][libm][GPU] Add missing vendor entrypoints to the GPU version of `libm`

AntonRydahl opened the pull request:
There are a number of entry points in `libm` for which no target-agnostic implementations exist, and the built-ins do not lower correctly on either AMDGPU or NVPTX targets. This patch adds the vendor entry points as a temporary solution.

AntonRydahl committed:
[libc][libm][GPU] Add missing vendor entrypoints to the GPU version of `libm`

There are a number of entry-points in `libm` for which no target-agnostic implementations exist and the built-ins do not lower correctly on either AMDGPU or NVPTX targets. This patch adds the vendor entry-points as a temporary solution.

arsenm commented on libc/src/math/gpu/vendor/amdgpu/declarations.h:
```diff
@@ -15,35 +15,61 @@ namespace __llvm_libc {
-double __ocml_rint_f64(double);
 float __ocml_rint_f32(float);
-double __ocml_round_f64(double);
+double __ocml_rint_f64(double);
```
Always use llvm.rint

arsenm commented on libc/src/math/gpu/vendor/amdgpu/declarations.h:
```diff
@@ -15,35 +15,61 @@ namespace __llvm_libc {
-double __ocml_round_f64(double);
+double __ocml_rint_f64(double);
 float __ocml_round_f32(float);
+double __ocml_round_f64(double);
```
Always use llvm.round

arsenm commented on libc/src/math/gpu/vendor/amdgpu/declarations.h:
```diff
@@ -15,35 +15,61 @@ namespace __llvm_libc {
+double __ocml_log10_f64(double);
+float __ocml_log1p_f32(float);
+double __ocml_log1p_f64(double);
+float __ocml_log2_f32(float);
```
Use llvm.log, log2 and log10 for f32 do not use for f64

AntonRydahl replied:
Thanks a lot, Matt! I have replaced the 32-bit versions with LLVM intrinsics, as suggested.

arsenm commented on libc/src/math/gpu/vendor/amdgpu/declarations.h:
```diff
@@ -15,35 +15,61 @@ namespace __llvm_libc {
+double __ocml_log2_f64(double);
+float __ocml_log_f32(float);
+double __ocml_log_f64(double);
+long __ocml_lrint_f32(float);
```
Should just use llvm.lrint for these, not sure if they work right now though. I think this is just llvm.rint plus cast

AntonRydahl replied:
I just looked more carefully through my test results. I think only the target agnostic implementation from `libc/src/math/generic` produces correct results on AMD GPU targets. __nv_rint/__nv_rintf is correct on NVIDIA targets, though.

I think it would be better to remove them for now and rely on `libc/src/math/generic/rint.cpp`, `libc/src/math/generic/lrint.cpp`, and `libc/src/math/generic/llrint.cpp`.

arsenm replied:
llvm.rint/lrint/llrint is the correct implementation

arsenm replied:
The ocml functions are pointless wrappers around the intrinsics. I wouldn't be shocked if llvm.lrint/llrint were broken in some way, but llvm.rint directly maps to an instruction and should work. Even if the instruction were broken, the correct implementation is have the backend work around the instruction

AntonRydahl replied:
That is a good point! I have replaced the vendor versions of llrint, llrintf, lrint, lrintf, llround, llroundf, lround, and lroundf with the corresponding intrinsics now.

AntonRydahl committed:
Replacing vendor rint and round functions with corresponding LLVM intrinsics

AntonRydahl committed:
Replaced vendor logf, log10f, and log2f with corresponding LLVM intrinsics in GPU libm

AntonRydahl commented:
@arsenm @jhuber6 would one of you have time to revisit this PR? I am sorry for spamming you!

jhuber6 commented:
Does it build and pass all the tests? The vendor stuff is straightforward since we don't test that and we'll delete it eventually anway. The `__builitin` implementations are more questionable since they're supposed to be the canonical implementation. (We still need to implement brute for math testing on the GPU, that's on the list).

AntonRydahl commented:
Sorry for the late reply! I passes all test on AMD GPUs at least

AntonRydahl committed:
Merged main into libc_vendor_functions

jhuber6 commented:
Pretty sure this all needs to be rebased on the recent `LIBC_NAMESPACE` change.

AntonRydahl committed:
Merged main into libc_vendor_functions

AntonRydahl committed:
Merge branch 'main' into libc_vendor_functions

AntonRydahl committed:
Updated namespace __llvm_libc to namespace LIBC_NAMESPACE

The change could be done with this command:
```
for file in `grep -rn libc/src/math -e "namespace.*libc" -l`; do echo "updating $file"; sed -i -e 's/__llvm_libc/LIBC_NAMESPACE/g' $file; done
```

AntonRydahl committed:
Passing all tests on gfx906 and sm_60 with and without vendor math enabled

AntonRydahl committed:
Removing __builtin_logf and __builtin_log2f and unused files

AntonRydahl committed:
Merge branch 'llvm:main' into libc_vendor_functions

jhuber6 commented on libc/src/math/gpu/lroundf.cpp:
```diff
@@ -0,0 +1,17 @@
+//===-- Implementation of the GPU lroundf function
```
This and some other headers need to be reformatted.

jhuber6 commented on libc/src/math/gpu/vendor/atan2f.cpp:
```diff
@@ -0,0 +1,21 @@
+//===-- Implementation of the GPU atan2f function
```
Fix header

jhuber6 commented on libc/src/math/gpu/vendor/logbf.cpp:
```diff
@@ -0,0 +1,19 @@
+//===-- Implementation of the GPU logbf function
```
Fix header

jhuber6 reviewed (approved):
LG after fixing the header formatting.

AntonRydahl committed:
Updated misformatted headers

AntonRydahl commented:
> LG after fixing the header formatting.

Thanks a bunch! I don't know how I did not notice the misformatted headers. I have fixed those now.
